cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,613 @@
+					objPath: {
+						segments:       segments,
+						hasLastSegment: false,
+					},
+				},
+			}
+		}
+
+		var (
+			buf   = &bytes.Buffer{}
+			obsvr = observer{
+				db:      teststore.New(),
+				writer:  csv.NewWriter(buf),
+				objects: bucketsObjs,
+			}
+		)
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		err := obsvr.analyzeProject(ctx.Context)
+		require.NoError(t, err)
+
+		// TODO: Add assertions for buf content
+	})
+
+	t.Run("object with non sequenced segments", func(t *testing.T) {
+		var bucketsObjs bucketsObjects
+		{ // Generate an objects without last segment
+			const (
+				bucketName = "analyzeBucket"
+				objPath    = storj.Path("analyzePath")
+			)
+			var segments uint64
+			{ // Calculate a unaligned number of segments
+				segments = rand.Uint64()
+				for {
+					trailingZeros := bits.TrailingZeros64(segments)
+					leadingZeros := bits.LeadingZeros64(segments)
+
+					if (trailingZeros + leadingZeros) == 64 {
+						continue
+					}
+
+					ones := bits.OnesCount64(segments)
+					if (trailingZeros + leadingZeros + ones) == 64 {
+						continue
+					}
+
+					break
+				}
+			}
+
+			bucketsObjs = bucketsObjects{
+				bucketName: map[storj.Path]*object{
+					objPath: {
+						segments:       segments,
+						hasLastSegment: true,
+					},
+				},
+			}
+		}
+
+		var (
+			buf   = &bytes.Buffer{}
+			obsvr = observer{
+				db:      teststore.New(),
+				writer:  csv.NewWriter(buf),
+				objects: bucketsObjs,
+			}
+		)
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		err := obsvr.analyzeProject(ctx.Context)
+		require.NoError(t, err)
+
+		// TODO: Add assertions for buf content
+	})
+
+	t.Run("object with unencrypted segments with different stored number", func(t *testing.T) {
+		var bucketsObjs bucketsObjects
+		{ // Generate an object
+			const (
+				bucketName = "analyzeBucket"
+				objPath    = storj.Path("analyzePath")
+			)
+			var (
+				segments           uint64
+				invalidNumSegments byte
+			)
+			{
+				numSegments := rand.Intn(62) + 1
+				segments = math.MaxUint64 >> numSegments
+
+				for {
+					numSeg := rand.Intn(65)
+					if numSeg != numSegments {
+						invalidNumSegments = byte(numSeg)
+						break
+					}
+				}
+			}
+
+			bucketsObjs = bucketsObjects{
+				bucketName: map[storj.Path]*object{
+					objPath: {
+						segments:                 segments,
+						expectedNumberOfSegments: invalidNumSegments,
+						hasLastSegment:           false,
+					},
+				},
+			}
+		}
+
+		var (
+			buf   = &bytes.Buffer{}
+			obsvr = observer{
+				db:      teststore.New(),
+				writer:  csv.NewWriter(buf),
+				objects: bucketsObjs,
+			}
+		)
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		err := obsvr.analyzeProject(ctx.Context)
+		require.NoError(t, err)
+
+		// TODO: Add assertions for buf content
+	})
+}
+
+// objectSegmentRef is an object segment reference to be used for simulating
+// calls observer.processSegment
+type objectSegmentRef struct {
```

Comment:
Can we have this type name as `segmentRef`? It's clear that segments belong to objects.

Reply:
done